calculator_test.go

```diff
@@ -2,23 +2,313 @@ package calculator_test
+		}
+	}
+
+}
+
+func TestAddRandom(t *testing.T) {
+	t.Parallel()
+	rand.Seed(time.Now().UnixNano())
+	for i := 0; i < 10; i++ {
+		a := rand.Float64()
+		b := rand.Float64()
+		want := a + b
+		got := calculator.Add(a, b)
+		if want != got {
+			t.Fatalf("Add(%f, %f): want %f, got %f", a, b, want, got)
+		}
+	}
+}
+
+func TestSubtract(t *testing.T) {
+	t.Parallel()
+	testCases := []struct {
+		name string
+		a, b float64
+		want float64
+	}{
+		{
+			name: "Six minus five point two equals zero point eight",
+			a:    6,
+			b:    5.2,
+			want: 0.8,
+		},
+		{
+			name: "Negative two minus four equals -negative six",
+			a:    -2,
+			b:    4,
+			want: -6,
+		},
+		{
+			name: "Five minus five equals zero",
+			a:    5,
+			b:    5,
+			want: 0,
+		},
+	}
+
+	for _, tc := range testCases {
+		got := calculator.Subtract(tc.a, tc.b)
+		if !CloseEnough(tc.want, got) {
+			t.Errorf("%s : Add(%f , %f): want %f, got %f", tc.name, tc.a, tc.b, tc.want, got)
+		}
+
+	}
+}
+
+func TestSubtractRandom(t *testing.T) {
+	t.Parallel()
+	rand.Seed(time.Now().UnixNano())
+	for i := 0; i < 10; i++ {
+		a := rand.Float64()
+		b := rand.Float64()
+		want := a - b
+		got := calculator.Subtract(a, b)
+		if want != got {
+			t.Fatalf("Subtract(%f, %f): want %f, got %f", a, b, want, got)
+		}
+	}
+}
+
+func TestMultiply(t *testing.T) {
+	t.Parallel()
+	testCases := []struct {
+		name    string
+		a, b    float64
+		want    float64
+		message string
+	}{
+		{
+			name: "Seven multiplied by six is forty two",
+			a:    7,
+			b:    6,
+			want: 42,
+		},
+		{
+			name: "Six point five multiplied by two is thirteen",
+			a:    6.5,
+			b:    2,
+			want: 13,
+		},
+		{
+			name: "Zero multiplied by three is zero",
+			a:    0,
+			b:    3,
+			want: 0,
+		},
+	}
+
+	for _, tc := range testCases {
+		got := calculator.Multiply(tc.a, tc.b)
+		if tc.want != got {
+			t.Errorf("%s : Multiply(%f , %f): want %f, got %f", tc.name, tc.a, tc.b, tc.want, got)
+		}
+	}
+}
+
+func TestMultiplyRandom(t *testing.T) {
+	t.Parallel()
+	rand.Seed(time.Now().UnixNano())
+	for i := 0; i < 10; i++ {
+		a := rand.Float64()
+		b := rand.Float64()
+		want := a * b
+		got := calculator.Multiply(a, b)
+		if want != got {
+			t.Fatalf("Multiply(%f, %f): want %f, got %f", a, b, want, got)
+		}
 	}
 }
 
-// func TestSubtract(t *testing.T) {
-// 	t.Parallel()
-// 	var want float64 = 2
-// 	got := calculator.Subtract(4, 2)
-// 	if want != got {
-// 		t.Errorf("want %f, got %f", want, got)
-// 	}
-// }
+func CloseEnough(a, b float64) bool {
+	tolerance := 0.0000002
+	return math.Abs(a-b) <= tolerance
+}
+
+func TestDivide(t *testing.T) {
+	t.Parallel()
+	testCases := []struct {
+		name        string
+		a, b        float64
+		want        float64
+		errExpected bool
+	}{
+		{
+			name:        "Seven divided by two is three point five",
+			a:           7,
+			b:           2,
+			want:        3.5,
+			errExpected: false,
+		},
+		{
+			name:        "Six divided by zero is nine hundred and ninety nine",
```

Comment:
But it isn't. We might name this test case something like 'Division by zero is an error'.